cmd/gapit/export_replay.go

```diff
@@ -98,5 +127,293 @@ func (verb *exportReplayVerb) Run(ctx context.Context, flags flag.FlagSet) error
+			return fmt.Errorf("Cannot find any directory under " + toolsPathParent)
+		}
+		sort.Strings(matches)
+		toolsPath := matches[len(matches)-1]
+
+		aapt := path.Join(toolsPath, "aapt")
+		zipalign := path.Join(toolsPath, "zipalign")
+		apksigner := path.Join(toolsPath, "apksigner")
+
+		tmpAPK := "tmp.apk"
+
+		// Re-assemble new APK with the new manifest
+		baseJar := path.Join(sdkPath, "platforms", "android-"+strconv.Itoa(targetSdkVersion), "android.jar")
+		if _, err := os.Stat(verb.Apk); err == nil {
+			return fmt.Errorf("Cannot find android platform %d, please install it.", targetSdkVersion)
+		}
+
+		if err := execCommand(ctx, aapt, "package", "-f", "-M", "AndroidManifest.xml", "-I", baseJar, "-F", tmpAPK); err != nil {
+			return err
+		}
+
+		// Add replay assets, uncompressed
+		assets, err := ioutil.ReadDir(assetsPath)
+		if err != nil {
+			return err
+		}
+		for _, a := range assets {
+			// Arguments ("-0", "") ensure the asset will not be compressed
+			if err := execCommand(ctx, aapt, "add", "-0", "", tmpAPK, path.Join(assetsPath, a.Name())); err != nil {
+				return err
+			}
+		}
+
+		// Add the replay libraries
+		// TODO(#2677): get rid of this workaround once ABI naming is sanitized
+		abi := capture.ABI.Name
+		if abi == "armeabi" {
+			abi = "armeabi-v7a"
+		}
+
+		files := []string{
+			"classes.dex",
+			path.Join("lib", abi, "libgapir.so"),
+			path.Join("lib", abi, "libVkLayer_VirtualSwapchain.so"),
+		}
+
+		for _, f := range files {
+			if err := execCommand(ctx, aapt, "add", "-0", "", tmpAPK, f); err != nil {
+				return err
+			}
+		}
+
+		// Zip-align, output in final APK file
+		replayAPKPath := path.Join(startdir, replayAPK)
+		if err := execCommand(ctx, zipalign, "4", tmpAPK, replayAPKPath); err != nil {
+			return err
+		}
+
+		// Sign the new APK
+		keystorePath := path.Join(os.ExpandEnv("${HOME}"), ".android", "debug.keystore")
+		if _, err := os.Stat(keystorePath); err != nil {
+			// No keystore found, create one
+			keystorePath = "debug.keystore"
+			// https://developer.android.com/studio/publish/app-signing#debug-mode
+			keytool := "keytool"
+			if _, err := exec.LookPath("keytool"); err != nil {
+				// keytool is not found in PATH, look in JAVA_HOME/bin
+				keytool = path.Join(os.ExpandEnv("JAVA_HOME"), "bin")
+				if _, err := os.Stat(keytool); err != nil {
+					return fmt.Errorf("Cannot find the 'keytool' command")
+				}
+			}
+			if err := execCommand(ctx, keytool, "-genkey", "-dname", "CN=Android Debug,O=Android,C=US", "-v", "-keystore", keystorePath, "-storepass", "android", "-alias", "androiddebugkey", "-keypass", "android", "-keyalg", "RSA", "-keysize", "2048", "-validity", "10000"); err != nil {
+				return err
+			}
+		}
+		if err := execCommand(ctx, apksigner, "sign", "--ks", keystorePath, "--ks-pass", "pass:android", replayAPKPath); err != nil {
+			return err
+		}
+
+	}
+
 	return nil
 }
+
+func execCommand(ctx context.Context, name string, args ...string) error {
```

Comment:
We have core/os/shell that provides a nice API to do this sort of thing.

Reply:
thanks for the pointer